src/services/oss/dir_stream.rs

```diff
@@ -0,0 +1,254 @@
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use bytes::Buf;
+use quick_xml::de;
+use serde::Deserialize;
+use time::format_description::well_known::Rfc3339;
+use time::OffsetDateTime;
+
+use super::backend::Backend;
+use super::error::parse_error;
+use crate::error::new_other_object_error;
+use crate::http_util::parse_error_response;
+use crate::object::ObjectPageStream;
+use crate::ops::Operation;
+use crate::path::build_rel_path;
+use crate::ObjectEntry;
+use crate::ObjectMetadata;
+use crate::ObjectMode;
+
+pub struct DirStream {
+    backend: Arc<Backend>,
+    root: String,
+    path: String,
+
+    token: String,
+
+    done: bool,
+}
+
+impl DirStream {
+    pub fn new(backend: Arc<Backend>, root: &str, path: &str) -> Self {
+        Self {
+            backend,
+            root: root.to_string(),
+            path: path.to_string(),
+
+            token: "".to_string(),
+
+            done: false,
+        }
+    }
+}
+
+#[async_trait]
+impl ObjectPageStream for DirStream {
+    async fn next_page(&mut self) -> Result<Option<Vec<ObjectEntry>>> {
+        if self.done {
+            return Ok(None);
+        }
+
+        let resp = self
+            .backend
+            .list_object(&self.path, self.token.clone())
+            .await?;
+
+        if resp.status() != http::StatusCode::OK {
+            if resp.status() == http::StatusCode::NOT_FOUND {
+                self.done = true;
+                return Ok(None);
+            }
+            let er = parse_error_response(resp).await?;
+            let err = parse_error(Operation::List, &self.path, er);
+            return Err(err);
+        }
+
+        let bs = resp.into_body().bytes().await.map_err(|e| {
+            new_other_object_error(Operation::List, &self.path, anyhow!("read body: {:?}", e))
+        })?;
+
+        let output: ListBucketOutput = de::from_reader(bs.reader()).map_err(|e| {
+            new_other_object_error(
+                Operation::List,
+                &self.path,
+                anyhow!("deserialize list_bucket output: {:?}", e),
+            )
+        })?;
+
+        self.done = !output.is_truncated;
+        self.token = output.next_continuation_token.clone().unwrap_or_default();
+
+        let mut entries = Vec::with_capacity(output.common_prefixes.len() + output.contents.len());
+
+        for prefix in output.common_prefixes {
+            let de = ObjectEntry::new(
+                self.backend.clone(),
+                &build_rel_path(&self.root, &prefix.prefix),
+                ObjectMetadata::new(ObjectMode::DIR),
+            )
+            .with_complete();
+            entries.push(de);
+        }
+
+        for object in output.contents {
+            if object.key.ends_with('/') {
+                continue;
+            }
+            let mut meta = ObjectMetadata::new(ObjectMode::FILE);
+
+            meta.set_etag(&object.etag);
+            meta.set_content_md5(object.etag.trim_matches('"'));
```

Comment:
Please make sure `head_object` returns more meta. If head_object returns the same meta here, we can set `with_complete()`.